### PR TITLE
Add grader support for runes

### LINKED ISSUE #9: Add support for runes library
In general,

```
runes ----(functions)----> modified_runes
```

I see two ways of grading runes.

### Checking the end product (modified_runes)
- Use something like [headless-gl](https://github.com/stackgl/headless-gl) to provide a webGL api for the AWS lambda node environment.
- Expose this library to the grader program
- The grader program is then able to check buffers generated by the student programs versus whatever they want, and award marks accordingly

**Pros**
- No magic, staff have control over what is happening, since the comparison happens in the grader programs
- Grade API remains largely unchanged and therefore consistent among assessments

**Cons**
- Not an elegant solution. There is also the chance that some rune functions are not commutative, i.e. perhaps `quarter_turn_right(my_rune)` and `quarter_turn_left(quarter_turn_left(quarter_turn_left(my_rune)))` do not generate the same end product
- Staff must learn and manage another api
- Not an elegant solution

### Checking the steps (functions)
- Using something like [acorn](https://github.com/acornjs/acorn), which is already a sub-dependency of js-slang, we generate the AST of the student program and parse it into a series of function calls.
- The grader simplifies this series of function calls according to a set of rules that we define here
- The grader program provides an example of a solution; this to is simplified
- The simplified student and grader series of function calls are compared. Marks are awarded if they match.

Below, `const right = quarter_turn_right; const left = quarter_turn_left;`. For example, the student writes,

```javascript
function upside_down(rune) {
    const theRune = rune;
    const runeTurned = right(theRune);
    const huh = left(left(left(runeTurned)));
    return huh
}
```

And the grader program is,

```javascript
const upside_down = rune => right(right(rune));
```

We define the set of rules,
- Compare functions of the same name
- Of these functions, compare their return values
- Of the return values, only our defined rune functions, and the function parameter is allowed
- `(r) => left(r)` is the same as `(r) => right(right(right(r)))`

Then, the grader program compares:

```javascript
// Compare functions of the same name
STUDENT: upside_down
GRADER: upside_down
// Compare their return values
STUDENT: huh
GRADER: right(right(rune))
// Only defined rune functions and function parameters allowed (resolve references)
STUDENT: left(left(left(runeTurned)))
// Only defined rune functions and function parameters allowed (resolve references)
STUDENT: left(left(left(right(theRune))))
// Only defined rune functions and function parameters allowed (resolve references)
STUDENT: left(left(left(right(rune))))
// `(r) => left(r)` is the same as `(r) => right(right(right(r)))`
STUDENT: right(right(rune))
GRADER: right(right(rune))
// marks awarded
```

**Pros**
- Much more elegant of a solution
- Does not depend on the behaviour of whatever node webgl substitute we use

**Cons**
- This process is abstracted away from the staff, and so they must learn and follow an API
- The grader API will be different for runes than other assessments, since the graderProgram is now not a function that returns a mark, but a function that demonstrates the correct solution
- In addition, the XML element `GRADERPROGRAM` will be required to contain a new attribute that is the mark awarded if this process completes with a match between student and grader programs
- Must spend quite a lot of time defining rules

### PR DESCRIPTION
**Pull request by thenaesh**
Partially fixes #9 

Support for runes is currently prone to occasional false negatives (student is correct but the autograder says it is wrong) when the picture is generated in technically correct but convoluted ways. In particular, `quarter_turn_right(quarter_turn_right(stack(quarter_turn_left(nova_bb), quarter_turn_left(nova_bb))))` should be identical to `stack(quarter_turn_right(nova_bb), quarter_turn_right(nova_bb))` but is misidentified as different by the autograder.

Unfortunately, fixing this requires a computer graphics specialist (which I am not), as most of the rune code is hand-written WebGL. I've identified the problem as some values being negative (when they should be positive) in the `instanceArray` of the flattened shape, but am unable to make sense of that.

In the meantime, the what we can do is to ask students to sound off if their programs have been falsely marked as incorrect. This way, the TA's workload only increases by a bit (when a student sounds off, then manually regrade their program). Since most of the rune missions are such that there is only one reasonable way to solve the problem, I think we shouldn't run into too many of such cases.

### CODE REVIEW

**Comment by thenaesh** [comment]
Also... test coverage for `src/graphics/rune_library.js` is not happening. That's a frankenstein copy-paste job from the library files in cadet-frontend, and I don't think much of the code (except the matrix library) is particularly testable.

Is there a way to exclude that file?

File: package.json [hunk 1]
```diff
@@ -4,14 +4,15 @@
   "main": "index.js",
   "license": "MIT",
   "scripts": {
-    "build": "rm -rf node_modules && yarn install --production=true && tsc && zip -r --exclude=*terraform* ./grader.zip index.js node_modules/ && yarn install --production=false",
+    "build": "rm -rf node_modules && rm -rf build && yarn install --production=true && tsc && cd build && zip -r --exclude=*terraform* ../dist/grader.zip index.js node_modules/ && cd .. && yarn install --production=false",
```
Comment (ningyuansg) [inline package.json]:
```        zip warning: name not matched: node_modules/
zip I/O error: No such file or directory
zip error: Could not create output file (../dist/grader.zip)
```

This is big enough to be its own bash script. Could you make one and place it in `./scripts`? For reference, the [scripts directory over at frontend](https://github.com/source-academy/cadet-frontend/tree/master/scripts).

File: package.json [hunk 2]
```diff
@@ -4,14 +4,15 @@
...
     "typecheck": "tsc --noEmit",
-    "test": "TIMEOUT=1000 jest",
-    "test-coveralls": "TIMEOUT=1000 jest --coverage --coverageReporters=text-lcov | coveralls"
+    "test": "TIMEOUT=1000 jest --env=node",
+    "test-coveralls": "TIMEOUT=1000 jest --env=node --coverage --coverageReporters=text-lcov | coveralls"
   },
   "dependencies": {
     "@types/node": "^10.5.2",
     "js-slang": "0.1.4",
+    "lodash": "4.17.10",
```
Comment (ningyuansg) [inline package.json]: I can't find lodash being used in this change

File: src/__tests__/examples/runes.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
+import { Grader, Student } from './types';
+
+const validStudentCorrect =
+    `
+    const a = () => blue(quarter_turn_left(quarter_turn_left(quarter_turn_left(nova_bb))));
+    //const b = () => quarter_turn_right(quarter_turn_right(stack(quarter_turn_left(nova_bb), quarter_turn_left(nova_bb))));
```
Comment (ningyuansg) [inline src/__tests__/examples/runes.ts]: Probably remove this commented line of code

File: src/index.ts [hunk 4]
```diff
@@ -2,6 +2,7 @@ import { createContext, runInContext, Result as SourceResult } from 'js-slang'
 import { SourceError } from 'js-slang/dist/types'
 
 const TIMEOUT_DURATION = parseInt(process.env.TIMEOUT!, 10) // in milliseconds
+require('./graphics/rune_library')
```
Comment (ningyuansg) [inline src/index.ts]: Prefer ES6 `import`. Also, would prefer a function `importExternals` that uses a switch to decide what libs to import. Probably place this function call in runAll before the call to evaluateGlobals. Would prevent name conflicts when more external libs are added.

**Comment by ningyuansg** [comment]
Bump

File: package.json (hunk 2 above)
Comment (thenaesh) [reply package.json]: It's in `src/graphics/rune_library.js`, specifically in the code that cleans up the flattened image data (`__compile`).

File: src/index.ts (hunk 4 above)
Comment (remo5000) [reply src/index.ts]: For `.js` files, I used `require()` as well, and kept a single object/function as the default export. If `import` works here, that's even better

File: package.json (hunk 2 above)
Comment (remo5000) [reply package.json]: Not sure what's the magnitude change in size, but if you're using one function, individual lodash function imports are available

Comment (thenaesh) [reply package.json]: Here's the exact code:

```
function __compile(shape) {
  var shape_list = generateFlattenedShapeList(shape);
  var cleaned_shape_list = shape_list.map(function (obj) {
      var shape = obj.shape;
      var instanceArray = _.mapValues(obj.instanceArray, function (value) {
          return parseFloat(value.toFixed(3));
      });
      return {
          shape: shape,
          instanceArray: instanceArray
      };
  });
  return cleaned_shape_list;
}

function __are_pictures_equal(shape0, shape1) {
    var compiled_shape0 = __compile(shape0);
    var compiled_shape1 = __compile(shape1);
    return _.isEqual(compiled_shape0, compiled_shape1);
}
```

So it's more than one function, and anyway this library is pretty general and useful just about everywhere.

File: scripts/build.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+fi
+
+if [ -f grader.zip ]
+    then
+        rm grader.zip
+fi
+
+yarn install --production=true
+tsc
+cd build && zip -r --exclude=*terraform* ../grader.zip index.js graphics/ && cd ..
```
Comment (ningyuansg) [inline scripts/build.sh]: Can't find `./graphics/*` in the zip file, may need a `-r` flag somewhere. Please verify the built zip file with `$ unzip -l grader.zip`.

Comment (thenaesh) [reply scripts/build.sh]: Strange... I checked the zip file on my machine and it contains `./graphics/*`. I'll have a look again.

Comment (thenaesh) [reply scripts/build.sh]: Here's the output of a freshly-built zip file with the node modules truncated:

```
Archive:  grader.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
     3623  2018-08-19 23:09   index.js
        0  2018-08-19 23:09   graphics/
   122764  2018-08-19 23:09   graphics/rune_library.js
   140554  2018-08-19 23:09   graphics/rune_library.js.map
        0  2018-08-19 23:09   node_modules/
        0  2018-08-19 23:09   node_modules/acorn/
    11096  2018-08-15 20:59   node_modules/acorn/CHANGELOG.md
        0  2018-08-19 23:09   node_modules/acorn/dist/
    47836  2018-08-15 20:59   node_modules/acorn/dist/acorn_loose.es.js
   179412  2018-08-15 20:59   node_modules/acorn/dist/acorn.js
    14926  2018-08-15 20:59   node_modules/acorn/dist/walk.js
    49813  2018-08-15 20:59   node_modules/acorn/dist/acorn_loose.js
        0  2018-08-15 20:59   node_modules/acorn/dist/.keep
...
        0  2018-08-19 23:09   node_modules/typescript/lib/tr/
   159296  2018-08-19 12:19   node_modules/typescript/lib/tr/diagnosticMessages.generated.json
     1103  2018-08-19 12:19   node_modules/typescript/lib/watchGuard.js
    10300  2018-08-19 12:19   node_modules/typescript/lib/lib.es2015.symbol.wellknown.d.ts
     9197  2018-08-19 12:19   node_modules/typescript/LICENSE.txt
     4965  2018-08-19 12:19   node_modules/typescript/README.md
---------                     -------
 41411914                     1525 files

```

The graphics folder is there together with its contents.

Comment (ningyuansg) [reply scripts/build.sh]:
```
ning:~/github/grader $ yarn build > /dev/null && git show --summary && unzip -l grader.zip | head
warning " > ts-jest@23.0.0" has unmet peer dependency "babel-core@^6.26.3 || ^7.0.0-0".              
warning " > ts-jest@23.0.0" has unmet peer dependency "babel-core@^6.26.3 || ^7.0.0-0".              
commit a3b66db41285e4acf590390a8b8495747d4cabb7 (HEAD -> support-runes, origin/support-runes)
Author: Thenaesh Elango <thenaeshelango@gmail.com>
Date:   Sun Aug 19 21:45:11 2018 +0800

    Ignore grader.zip

Archive:  grader.zip
  Length      Date    Time    Name
---------  ---------- -----   ----
     3683  2018-08-20 00:05   index.js
        0  2018-08-20 00:05   node_modules/
        0  2018-08-20 00:05   node_modules/js-slang/
        0  2018-08-20 00:05   node_modules/js-slang/dist/
      868  2018-08-12 17:03   node_modules/js-slang/dist/interpreter.d.ts
     1635  2018-08-12 17:03   node_modules/js-slang/dist/types.js
    16768  2018-08-12 17:03   node_modules/js-slang/dist/interpreter.js.map
```

Alright, something weird is going on here

Comment (ningyuansg) [reply scripts/build.sh]: Oh, so I think that you are zipping in a `./graphics` directory that was generated *previously* by tsc, probably when the target directory for `tsc` was still the project root. I never ran `tsc` when that was the case, so I don't have the `./graphics` directory (and it's not tracked by git either).

You should probably be zipping `./build/src/graphics` in there. Although, you will need to be careful about whether the `index.js` file can import the `graphics/*` libraries correctly (i.e. file path troubles). Would be best to unzip the `grader.zip` package somewhere and just run the `exports.runAll` function.

**Comment by ningyuansg** [comment]
@thenaesh Bump, mission 1 is due quite soon and im anxious
@remo5000 Can you take over the review? Pls make sure that a fresh repo compiles, and that the built index.js file runs properly (the runAll function, basically)
@indocomsoft Would be good to test this master branch after this PR is merged on staging, before the scheduled task for grading mission 1 goes out

**Comment by thenaesh** [comment]
@ningyuansg Unfortunately, due to a personal emergency, I'll be unable to work on this PR for a while. Could someone else take over? So sorry for the inconvenience.

**Comment by remo5000** [comment]
Sure thing. The files are moved as expected, and I managed to run the test code (for 2d runes) in the node environment:

image

The only things I had to do (which wouldn't be a problem in any other case) are 
- Setting a `TIMEOUT` variable for the last grader program
- defining `exports` in the node environment, as it treats the index.js file as a script

LGTM.

**Comment by remo5000** [comment]
Added some dummy visualisation functions, as it is commonplace for solutions to include the test functions (something that was expected in the last academic year, in fact)

File: README.rst [hunk 6]
```diff
@@ -23,6 +23,14 @@ The grader is a component of the `Cadet backend`_. The grader,
 .. _Cadet backend: https://github.com/source-academy/cadet
 .. _js-slang: https://github.com/source-academy/js-slang
 
+========================
+External Library Support
+========================
+
+Only support for the "runes" library is available. These include the functions:
+- `__compile`
```
Comment (ningyuansg) [inline README.rst]: @remo5000 Would be good to describe this function. It's not immediately clear what this does.

**Pull request merged**